dissect/target/plugins/apps/texteditor/windowsnotepad.py

```diff
@@ -0,0 +1,377 @@
+    uint8       rightToLeft;
+    uint8       showUnicode;
+    uint8       optionsVersion;
+};
+
+struct tab_header_crc32_stub {
+    char        unk1;
+    char        unk2;
+    char        crc32[4];
+};
+
+struct fixed_size_data_block {
+    uleb128     nAdded;
+    wchar       data[nAdded];
+    uint8       hasRemainingVariableDataBlocks; // indicates whether after this single-data block more data will follow
+    char        crc32[4];
+};
+
+struct variable_size_data_block {
+    uleb128     offset;
+    uleb128     nDeleted;
+    uleb128     nAdded;
+    wchar       data[nAdded];
+    char        crc32[4];
+};
+
+struct options_v1 {
+    uleb128     unk;
+};
+
+struct options_v2 {
+    uleb128     unk1; // likely autocorrect or spellcheck
+    uleb128     unk2; // likely autocorrect or spellcheck
+};
+"""
+
+WINDOWS_SAVED_TABS_EXTRA_FIELDS = [("datetime", "modification_time"), ("string", "sha256"), ("path", "saved_path")]
+
+
+class WindowsSavedTabRecordDescriptorExtension(RecordDescriptorExtensionBase):
+    """RecordDescriptorExtension used to add extra fields to tabs that are saved to disk and contain more info."""
+
+    _default_fields = WINDOWS_SAVED_TABS_EXTRA_FIELDS
+
+    _input_fields = ("_saved",)
+
+    def _fill_default_fields(self, record_kwargs):
+        r: WindowsNotepadSavedTabContentRecord = record_kwargs.get("_saved", None)
+
+        modification_time = None
+        saved_path = None
+        sha256 = None
+
+        if r:
+            modification_time = r.modification_time
+            sha256 = r.sha256
+            saved_path = r.saved_path
+
+        record_kwargs.update({"modification_time": modification_time, "sha256": sha256, "saved_path": saved_path})
+        return record_kwargs
+
+
+# Different Record types for both saved/unsaved tabs, and with/without UserRecordDescriptor so that the
+# plugin can be used as a standalone tool as well
+
+
+WindowsNotepadUnsavedTabRecord = create_extended_descriptor([UserRecordDescriptorExtension])(
+    "texteditor/windowsnotepad/tab/unsaved",
+    GENERIC_TAB_CONTENTS_RECORD_FIELDS,
+)
+
+WindowsNotepadSavedTabRecord = create_extended_descriptor(
+    [UserRecordDescriptorExtension, WindowsSavedTabRecordDescriptorExtension]
+)(
+    "texteditor/windowsnotepad/tab/saved",
+    GENERIC_TAB_CONTENTS_RECORD_FIELDS,
+)
+
+WindowsNotepadUnsavedTabContentRecord = create_extended_descriptor([])(
+    "texteditor/windowsnotepad/tab_content/unsaved", GENERIC_TAB_CONTENTS_RECORD_FIELDS
+)
+
+WindowsNotepadSavedTabContentRecord = create_extended_descriptor([])(
+    "texteditor/windowsnotepad/tab_content/saved", GENERIC_TAB_CONTENTS_RECORD_FIELDS + WINDOWS_SAVED_TABS_EXTRA_FIELDS
+)
+
+c_windowstab = cstruct()
+c_windowstab.load(c_def)
+
+
+def _calc_crc32(data: bytes) -> bytes:
+    """Perform a CRC32 checksum on the data and return it as bytes."""
+    return zlib.crc32(data).to_bytes(length=4, byteorder="big")
+
+
+class WindowsNotepadTabContent:
```

Comment:
It's a better idea to return a `WindowsNotepadTab` object, and leaving the creation of the records to the plugin code. This way, when I use this class to parse a Windows Notepad tab I have access to the low-level attributes like `tab. file_header.updateNumber` in my own code instead of getting just a `Record` representation of the tab.